app/upgrade_tracker.go

```diff
@@ -0,0 +1,95 @@
+package app
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+)
+
+type upgradeHandlerFn func(ctx sdk.Context, vm module.VersionMap) (module.VersionMap, error)
+
+type upgradeTrackerItem struct {
+	// Monotonically increasing index to order and track migrations. Typically the current unix epoch timestamp.
+	index int64
+	// Function that will run during the SetUpgradeHandler callback. The VersionMap must always be returned.
+	upgradeHandler upgradeHandlerFn
+	// StoreUpgrades that will be provided to UpgradeStoreLoader
+	storeUpgrade *storetypes.StoreUpgrades
+}
+
+// upgradeTracker allows us to track needed upgrades/migrations across both release and develop builds
+type upgradeTracker struct {
+	upgrades     []upgradeTrackerItem
+	stateFileDir string
+}
+
+func (t upgradeTracker) getDevelopUpgrades() ([]upgradeHandlerFn, *storetypes.StoreUpgrades, error) {
+	neededUpgrades := &storetypes.StoreUpgrades{}
+	neededUpgradeHandlers := []upgradeHandlerFn{}
+	stateFilePath := path.Join(t.stateFileDir, "developupgradetracker")
+
+	currentIndex := int64(0)
+	stateFileContents, err := os.ReadFile(stateFilePath) // #nosec G304 -- stateFilePath is not user controllable
+	if err == nil {
+		currentIndex, err = strconv.ParseInt(string(stateFileContents), 10, 64)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to decode upgrade tracker: %w", err)
```

Comment:
Can we add a unit test for this  using a sample.Developupgradetracker.

The sample file can be added to the testutil directory

Reply:
You're saying you want this specific branch to have test coverage? I can do that.